asserts/pool.go

```diff
@@ -19,6 +19,661 @@
+
+	return p.resolveWith(unresolved, uniq, u, a, extrag)
+}
+
+// TODO: AddBatch
+
+var (
+	ErrUnresolved       = errors.New("unresolved assertion")
+	ErrUnknownPoolGroup = errors.New("unknown pool group")
+)
+
+// unresolvedBookkeeping processes any left over unresolved assertions
+// since the last ToResolve invocation and intervening calls to Add/AddBatch,
+//  * they were either marked as in error which will be propagated
+//    to all groups requiring them
+//  * simply unresolved, which will be propagated to groups requiring them
+//    as ErrUnresolved
+//  * unchanged (update case)
+// unresolvedBookkeeping will also promote any recorded prerequisites
+// into actively unresolved, as long as not all the groups requiring them
+// are in error.
+func (p *Pool) unresolvedBookkeeping() {
+	// any left over unresolved are either:
+	//  * in error
+	//  * unchanged
+	//  * or unresolved
+	for uniq, u := range p.unresolved {
+		e := u.err
+		if e == nil {
+			if u.at.Revision == RevisionNotKnown {
+				e = ErrUnresolved
+			}
+		}
+		if e != nil {
+			p.setErr(&u.grouping, e)
+		}
+		delete(p.unresolved, uniq)
+	}
+
+	// prerequisites will become the new unresolved but drop them
+	// if all their groups are in error
+	for uniq, prereq := range p.prerequisites {
+		useful := false
+		p.groupings.Iter(&prereq.grouping, func(gnum uint16) error {
+			if !p.groups[gnum].hasErr() {
+				useful = true
+			}
+			return nil
+		})
+		if !useful {
+			delete(p.prerequisites, uniq)
+			continue
+		}
+	}
+
+	// prerequisites become the new unresolved, the emptied
+	// unresolved is used for prerequisites in the next round
+	p.unresolved, p.prerequisites = p.prerequisites, p.unresolved
+}
+
+// Err returns the error for group if group is in error, nil otherwise.
+func (p *Pool) Err(group string) error {
+	gnum, err := p.groupNum(group)
+	if err != nil {
+		return err
+	}
+	gRec := p.groups[gnum]
+	if gRec == nil {
+		return ErrUnknownPoolGroup
+	}
+	return gRec.err
+}
+
+// AddError associates error e with the unresolved assertion.
+// The error will be propagated to all the affected groups at
+// the next ToResolve.
+func (p *Pool) AddError(e error, ref *Ref) error {
+	if err := p.phase(poolPhaseAdd); err != nil {
+		return err
+	}
+	uniq := ref.Unique()
+	if u := p.unresolved[uniq]; u != nil && u.err == nil {
+		u.err = e
+	}
+	return nil
+}
+
+// AddGroupingError puts all the groups of grouping in error, with error e.
+func (p *Pool) AddGroupingError(e error, grouping Grouping) error {
+	if err := p.phase(poolPhaseAdd); err != nil {
+		return err
+	}
+
+	g, err := p.groupings.Parse(string(grouping))
+	if err != nil {
+		return err
+	}
+
+	p.setErr(g, e)
+	return nil
+}
+
+// AddToUpdate adds the assertion referenced by toUpdate and all its
+// prerequisites to the Pool as unresolved and as required by the
+// given group.
+// It is assumed that the assertion is currently in the ground
+// database of the Pool, otherwise this will error. These assertions
```

Comment:
The sentence "These assertions current revisions..." looks like it's missing something, and it sounds confusing.